### PR TITLE
Remove references to deprecated numpy globals

### PR DESCRIPTION
**Pull request by eric-wieser**
Fixes all the deprecations I introduced in numpy/numpy#14882 that can be found with `np\.(bool|int|float|complex|object|str|long|unicode)\b`.

For all these replacements, there are two choices. `np.float` can become either:

* `float`: identical meaning
* `np.float_`: different but likely intended meaning

As long as the result is passed to `np.dtype(...)` somewhere, there is no difference anyway.

For the tests I've leant towards the former, while for the rest of the code I've leant to the latter.

### CODE REVIEW

File: numba/core/types/__init__.py [hunk 1]
```diff
@@ -105,8 +105,8 @@
 uint = _make_unsigned(np.int_)
 intc = _make_signed(np.intc) # C-compat int
 uintc = _make_unsigned(np.uintc) # C-compat uint
-long_ = _make_signed(np.long)
-ulong = _make_unsigned(np.long)
+long_ = _make_signed(np.int_)  # C-compat long
+ulong = _make_unsigned(np.uint)  # C-compat ulong
```
Comment (eric-wieser) [inline numba/core/types/__init__.py:109]: Seems a bit silly that `np.int_` is here both as `uint` and `ulong`...

**Comment by gmarkall** [comment]
The CI failure is probably #5973.

**Review by gmarkall — APPROVED** [review]
The CUDA-related changes look good to me and test OK with hardware (2x RTX 8000) and cudasim.

**Comment by gmarkall** [comment]
//AzurePipelines run

**Comment by gmarkall** [comment]
@eric-wieser Thanks for the PR! I'll mark it as ready-for-review once we have got the CI to go through, but someone else will need to start it off (e.g. @stuartarchibald  @esc )

**Comment by gmarkall** [comment]
Marking as ready for review anyway - we can re-run CI later.

**Comment by eric-wieser** [comment]
Conflicts resolved I hope

**Comment by eric-wieser** [comment]
Has this missed the chance to make it into 0.52.0?

**Comment by gmarkall** [comment]
I'm popping this in the 0.52.0RC2 milestone for now as a placeholder to answer the question of whether it missed the chance for 0.52 - somehow it ended up not in a milestone, so I think it got missed for consideration so far.

**Comment by eric-wieser** [comment]
Should #6182 be in that milestone too?

**Comment by stuartarchibald** [comment]
> Should #6182 be in that milestone too?

What's the scheduled release date for NumPy 1.20 ?

**Comment by eric-wieser** [comment]
I'm actually not sure - it could be a way off yet. I only ask because there are two PRs mentioned in that issue; this one, and one that you already milestones and merged.

**Comment by sklam** [comment]
Given this is addressing deprecation from numpy1.20, a future version, and that there is a slight worry for windows being different, it's better to move it to 0.53.

**Comment by eric-wieser** [comment]
There's no concern of a behavior change on Windows here - `np.dtype(np.long)` and `np.dtype(np.int_)` are synonymous on all platforms.

**Comment by eric-wieser** [comment]
The future version argument is valid. It impacts me because I'm frequently running from a modified copy of numpy master, but that's likely not applicable to make other users.

**Review by stuartarchibald — COMMENTED** [review]
Thanks for making this update. I've checked this against NumPy 1.20.0RC2, `import numba` seems to not trigger anything which suggests `numba.types` is ok. I've also grepped around in the code base and run a random selection of tests to see if anything else is triggering this warning, seems ok once the few things noted in the review are addressed. Given 0.53 won't support 1.20 (as it's not released yet) I see this as getting ahead compatibility wise and if further adjustments are needed they can be made for 0.54 which I anticipate will support 1.20 officially.

RE this PR: a few minor thing to resolve and a build farm run to confirm and this is good to go I think. Thanks again!

File: numba/np/arrayobj.py [hunk 2]
```diff
@@ -3832,15 +3832,15 @@ def _arange_dtype(*args):
     else:
         # numerous attempts were made at guessing this type from the NumPy
         # source but it turns out on running `np.arange(10).dtype` on pretty
-        # much all platform and python combinations that it matched np.int?!
+        # much all platform and python combinations that it matched np.int_?!
```
[the remark this answers is not in the document]
Comment (eric-wieser) [reply numba/np/arrayobj.py:3835]: `np.arange(10).dtype == np.dtype(int) == np.int_` by definition, so perhaps this whole comment can go.

File: numba/core/types/__init__.py (hunk 1 above)
Comment (eric-wieser) [inline numba/core/types/__init__.py:109]:
```suggestion
ulong = _make_unsigned(np.int_)  # C-compat ulong
```

File: numba/np/arrayobj.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (eric-wieser) [reply numba/np/arrayobj.py]: I've updated the comment with a `?!`-less description.

**Comment by esc** [comment]
Build Farm ID: `numba_smoketest_cuda_yaml_5`.

**Comment by esc** [comment]
`numba_smoketest_cuda_yaml_5` passed.

**Comment by esc** [comment]
Build Farm ID: `numba_smoketest_cpu_yaml_1`

**Comment by esc** [comment]
`numba_smoketest_cpu_yaml_1` finally passed 🎉

**Review by esc — APPROVED** [review]
I've scanned through the code, I see that all reviewer comments have been adequately addressed and that this PR passed both the CUDA and the CPU smoketest on the Anaconda internal build farm. I conclude that all is well and this PR can and should be merged.

**Comment by esc** [comment]
@stuartarchibald @sklam @gmarkall since you were all involved in the review process, would you like to sign-off this one too?

**Review by stuartarchibald — APPROVED** [review]
Thanks for the fixes. Looks good.

**Pull request merged**